### ISSUE TITLE
Image.Thumbnail causes distortion on underlying file

### ISSUE THREAD

**Issue by majora2007** [kleisauke/net-vips#155]
I've been using NetVips successfully for many images, but a few have suddenly caused some weird distortions and I cannot understand why. 

This is the original file
![test expected](https://user-images.githubusercontent.com/735851/149230633-cac0eb24-1d14-4d01-ad05-77a1ac02ec94.jpg)

This is the thumbnail produced:
![test2_output](https://user-images.githubusercontent.com/735851/149230677-8beeb0dc-a9dc-4719-8d23-38f77e6b8408.png)

Here is the code
```
var expectedBytes = Image.Thumbnail("expected.jpg"), 320).WriteToBuffer(".png");
var thumbnail = Image.Thumbnail("input.jpg", 320);
thumbnail.WriteToFile("ouput/expected.png");
var actualBytes = File.ReadBytes("output/expected.png")

Assert.Equal(expectedBytes, actualBytes)
```

I tried both jpg and png (not sure if NetVips does anything different), but it always spits out this weird corrupted version. Yet windows viewer can open it fine. 

Any ideas?

**Comment by majora2007** [comment]
Sure, here is the original file:
https://drive.google.com/file/d/1JLCjnw-3uq42xXIpP0wNlb85dobWpwj5/view?usp=sharing

This is very interesting, could you also check on your PC maybe from Stream? 
(You can use this link or just zip up the image)
https://drive.google.com/file/d/1cYYHS91hAwuDUr-stVBUCy4SSSEBjGCk/view?usp=sharing

```
using var archive = ZipFile.OpenRead(archivePath);
var entry = archive.Entries.First();
using var stream = entry.Open();
using var thumbnail = Image.ThumbnailStream(stream, ThumbnailWidth);
thumbnail.WriteToFile("expected.png");
```

My original code is with a stream (DeflateStream) as I'm extracting from archives. But when I tested just the bare image, it was also failing. 

It's strange because nearly all other images (~20k) process fine, just a few in this group doesn't. Hence me wondering if there was some jpg compression in play here.

**Comment by kleisauke** [comment]
I could reproduce this. It looks like a streaming bug in libvips' JPEG loader.  Here's a reproduction in Python:
```python
import pyvips
from random import randrange

filename = "149230633-cac0eb24-1d14-4d01-ad05-77a1ac02ec94.jpg"

im = pyvips.Image.thumbnail(filename, 320)
buffer = im.write_to_buffer(".png")
buffer_length = len(buffer)

i = 0
while True:
    i += 1
    print(f"loop {i}")
    input_file = open(filename, "rb")

    def read_handler(size):
        new_size = randrange(size)
        #print(f"{size} -> {new_size}")
        return input_file.read(new_size)

    source = pyvips.SourceCustom()
    source.on_read(read_handler)

    im2 = pyvips.Image.thumbnail_source(source, 320);
    buffer2 = im2.write_to_buffer(".png")
    buffer2_length = len(buffer2)
    if buffer_length != buffer2_length:
        print(f"Buffer length not the same ({buffer_length} versus {buffer2_length}) -- this should not happen")
        break;
```
```bash
$ python sanity.py
loop 1
loop 2
loop 3
loop 4
loop 5
loop 6
Buffer length not the same (93602 versus 92456) -- this should not happen
```

It only seems to happen for streams that doesn't fill up the buffer at once (i.e. `bytes_read != len(buffer)`). /cc @jcupitt

As a possible workaround (for now), you could decompress the `DeflateStream` to a `MemoryStream`, for example:
```csharp
using var archive = ZipFile.OpenRead("test.expected.zip");
var entry = archive.Entries.First();
using var stream = entry.Open();
using var memoryStream = new MemoryStream();
stream.CopyTo(memoryStream);
memoryStream.Position = 0;

using var thumbnail = Image.ThumbnailStream(memoryStream, 320);
thumbnail.WriteToFile("x.png");
````

**Comment by jcupitt** [comment]
Wow ok that's bad :( 

I tried changing the save format, and it works with PPM images (?!?!?!?!) which is strange:

```python
#!/usr/bin/python3

import sys
import pyvips
from random import randrange

filename = sys.argv[1]
format = sys.argv[2]

im = pyvips.Image.thumbnail(filename, 320)
buffer = im.write_to_buffer(format)
buffer_length = len(buffer)
print(f"written {len(buffer)} bytes of {format}")

i = 0
while True:
    i += 1
    print(f"loop {i} ...")
    input_file = open(filename, "rb")

    def read_handler(size):
        new_size = randrange(size)
        #print(f"{size} -> {new_size}")
        return input_file.read(new_size)

    source = pyvips.SourceCustom()
    source.on_read(read_handler)

    im2 = pyvips.Image.thumbnail_source(source, 320);
    buffer2 = im2.write_to_buffer(format)

    if len(buffer) != len(buffer2):
        print(f"written {len(buffer2)} bytes ... this should not happen!")
        break;
```

I see:

```
$ ./corrupt.py ~/pics/bleach.jpg .ppm
written 194935 bytes of .ppm
loop 1 ...
loop 2 ...
loop 3 ...
loop 4 ...
loop 5 ...
loop 6 ...
loop 7 ...
loop 8 ...
loop 9 ...
loop 10 ...
loop 11 ...
loop 12 ...
loop 13 ...
loop 14 ...
loop 15 ...
...
```

But fails with most formats like JPG, PNG, WEBP, etc.

**Comment by jcupitt** [comment]
I think letting size go to zero (which could happen) is wrong -- that means EOF. Maybe:

```python
    def read_handler(size):
        return input_file.read(1 + randrange(size - 1))
```

is better (though still not working).

**Comment by kleisauke** [comment]
Indeed, it's strange. Could it be a libjpeg(-turbo) bug? I tried to simplify jpeg2vips slightly with commit https://github.com/kleisauke/libvips/commit/5b720ea61c44d54e1242359945a3997358b9156c (i.e. by using `vips_source_seek` in `skip_input_data`, and a few other simplifications), but that doesn't fix the issue.

I'll have a look at the libjpeg documentation.

**Comment by kleisauke** [comment]
I couldn't find anything obvious in the libjpeg documentation. However, I noticed that once a seek handler was attached, the bug could no longer be reproduced.

```diff
@ -23,8 +23,13 @@ while True:
             size = randrange(10, size)
         return input_file.read(size)
 
+    def seek_handler(offset, whence):
+        input_file.seek(offset, whence)
+        return input_file.tell()
+
     source = pyvips.SourceCustom()
     source.on_read(read_handler)
+    source.on_seek(seek_handler)
 
     im2 = pyvips.Image.thumbnail_source(source, 320)
     buffer2 = im2.write_to_buffer(format)
```

**Comment by kleisauke** [comment]
Commit https://github.com/kleisauke/libvips/commit/3ab0765e2a4fa46ea42209ca4c3c1cca9654a855 fixes this for me. It looks like the JPEG loader still needs the saved header bytes during the pixel decode phase for non-seekable sources.

**Comment by kleisauke** [comment]
FWIW, here is a deterministic variant of the above test program.

<details>
  <summary>Details</summary>
  
```python
#!/usr/bin/python3

import sys
import pyvips
import random
random.seed(42)

filename = sys.argv[1]
format = sys.argv[2]

im = pyvips.Image.thumbnail(filename, 320)
buffer = im.write_to_buffer(format)
print(f"written {len(buffer)} bytes of {format}")

i = 0
while True:
    i += 1
    print(f"loop {i} ...")
    input_file = open(filename, "rb")

    def read_handler(size):
        if size > 10:
            size = random.randrange(10, size)
        return input_file.read(size)

    source = pyvips.SourceCustom()
    source.on_read(read_handler)

    im2 = pyvips.Image.thumbnail_source(source, 320)
    buffer2 = im2.write_to_buffer(format)

    if len(buffer) != len(buffer2):
        print(f"written {len(buffer2)} bytes ... this should not happen!")
        break
```
</details>

**Comment by majora2007** [comment]
Glad to see that you've figured it out. Can I assume this will go out with the next NetVips release?

**Comment by jcupitt** [comment]
I don't think we've found the right fix yet, unfortunately :( I'm poking about now.

**Comment by jcupitt** [comment]
I think I may have found a fix, if you have time to test.

**Comment by jcupitt** [comment]
OK, tested and merged. This fix will be in libvips 8.12.2.
